cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeMonitor.java

```diff
@@ -159,11 +160,32 @@ private void publish(String topicConfig, MonitorConsumeRequest request,
     messagePublisher.publish(NamespaceId.SYSTEM.getNamespace(), cConf.get(topicConfig),
                              messages.stream().map(s -> s.getMessage().getBytes(StandardCharsets.UTF_8)).iterator());
 
-    topicsToRequest.put(topicConfig, new MonitorConsumeRequest(request.getTopic(), topicConfig,
-                                                               messages.get(messages.size() - 1).getMessageId(),
+    topicsToRequest.put(topicConfig, new MonitorConsumeRequest(messages.get(messages.size() - 1).getMessageId(),
                                                                limit));
   }
 
+  private void triggerRuntimeShutdown() throws Exception {
+    try {
+      restClient.execute(HttpRequest.builder(HttpMethod.POST, clientConfig.resolveURL("runtime/shutdown")).build());
+    } catch (ConnectException e) {
+      LOG.trace("Connection refused when attempting to connect to Runtime Http Server. " +
+                  "Assuming that it is not available.");
+    }
+  }
+
+  private void setIsRuntimeInactive(List<MonitorMessage> monitorMessages) {
+    for (MonitorMessage message : monitorMessages) {
+      Notification notification = GSON.fromJson(message.getMessage(), Notification.class);
+      String programStatus = notification.getProperties().get(ProgramOptionConstants.PROGRAM_STATUS);
```

Comment:
Please also open a JIRA / add to the existing security JIRA that the runtime monitor should check that the messages is actually intended for the program run ID that it is monitoring.